Ninja hides the exit status of failing subprocesses

In build.cc, Ninja checks if the subprocess succeeded or failed

```
  // Print the command that is spewing before printing its output.
  if (!success) {
    string outputs;
    for (vector<Node*>::const_iterator o = edge->outputs_.begin();
         o != edge->outputs_.end(); ++o)
      outputs += (*o)->path() + " ";

    printer_.PrintOnNewLine("FAILED: " + outputs + "\n");
    printer_.PrintOnNewLine(edge->EvaluateCommand() + "\n");
  }
```

But by this point in the build, Ninja has already discarded the actual exit value from the subprocess.

There are a variety of tools that don't print anything to the console when they fail to run, relying on the user to instead check the exit status to see what the reason for failure was.

This is especially problematic for third party tools where the source code is not available.

It's, of course, possible to get the exit status printed to the console using a wrapper script, but as discussed at extreme length, wrapper scripts are desired by basically no one.

Ninja should print the exit status of failing subprocesses.

I don't see any "FAILED:" message, but instead
```ninja: build stopped: subcommand failed```

Is it the same issue? 

@schultetwin1 Would https://github.com/ninja-build/ninja/pull/1805 address this situation too, and reveal the exit code of the failed subcommand?

@jonesmz @jimis @schultetwin1 have you guys figured this out?

My make file:
```
#!/bin/bash
# Build script
export CC=/usr/bin/clang
export CXX=/usr/bin/clang++
mkdir -p build
cd build
cmake .. -G "Ninja" -DCMAKE_BUILD_TYPE=Release
ninja
cd ..
```

Error in github actions (it still thinks the build is successful):
```
Run ./make
-- The C compiler identification is AppleClang 12.0.0.12000032
-- The CXX compiler identification is AppleClang 12.0.0.12000032
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /Applications/Xcode_12.2.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /Applications/Xcode_12.2.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Build spdlog: 1.8.2
-- Looking for pthread.h
-- Looking for pthread.h - found
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Success
-- Found Threads: TRUE  
-- Build type: Release
-- Configuring done
-- Generating done
-- Build files have been written to: /Users/runner/work/pvphero-launcher/pvphero-launcher/build
[1/20] Copying OS X Content PvPHero.app/Contents/Resources/icon.icns
[2/20] Building CXX object deps/spdlog/CMakeFiles/spdlog.dir/src/file_sinks.cpp.o
[3/20] Building CXX object deps/spdlog/CMakeFiles/spdlog.dir/src/stdout_sinks.cpp.o
[4/20] Building CXX object deps/spdlog/CMakeFiles/spdlog.dir/src/color_sinks.cpp.o
[5/20] Building CXX object deps/spdlog/CMakeFiles/spdlog.dir/src/spdlog.cpp.o
[6/20] Building CXX object CMakeFiles/PvPHero.dir/src/updating/Downloader.cpp.o
[7/20] Building CXX object CMakeFiles/PvPHero.dir/src/updating/client/ClientUpdater.cpp.o
[8/20] Building CXX object CMakeFiles/PvPHero.dir/src/PvPHero.cpp.o
[9/20] Building CXX object CMakeFiles/PvPHero.dir/src/external/sago/platform_folders.cpp.o
[10/20] Building CXX object CMakeFiles/PvPHero.dir/src/filesystem/InMemoryFileSystem.cpp.o
[11/20] Building CXX object CMakeFiles/PvPHero.dir/src/updating/Utils.cpp.o
../src/updating/Utils.cpp:61:39: warning: object backing the pointer will be destroyed at the end of the full-expression [-Wdangling-gsl]
        const char *command_line[] = {getJavaPath().c_str(), "-m", "pvphero/com.pvphero.PvPHero",
                                      ^~~~~~~~~~~~~
1 warning generated.
[12/20] Building C object CMakeFiles/PvPHero.dir/src/external/zip/zip.c.o
[13/20] Building CXX object CMakeFiles/PvPHero.dir/src/Assets.cpp.o
../src/Assets.cpp:2:9: warning: #pragma once in main file [-Wpragma-once-outside-header]
#pragma once
        ^
1 warning generated.
[14/20] Building CXX object deps/spdlog/CMakeFiles/spdlog.dir/src/async.cpp.o
[15/20] Building CXX object deps/spdlog/CMakeFiles/spdlog.dir/src/cfg.cpp.o
[16/20] Building CXX object CMakeFiles/PvPHero.dir/src/updating/launcher/LauncherUpdater.cpp.o
[17/20] Building CXX object deps/spdlog/CMakeFiles/spdlog.dir/src/fmt.cpp.o
[18/20] Linking CXX static library deps/spdlog/libspdlog.a
[19/20] Building CXX object CMakeFiles/PvPHero.dir/src/main.cpp.o
[20/20] Linking CXX executable PvPHero.app/Contents/MacOS/PvPHero
FAILED: PvPHero.app/Contents/MacOS/PvPHero 
: && /Applications/Xcode_12.2.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang++ -O3 -DNDEBUG -isysroot /Applications/Xcode_12.2.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX11.0.sdk -mmacosx-version-min=10.15 -Wl,-search_paths_first -Wl,-headerpad_max_install_names /SUBSYSTEM:CONSOLE -s CMakeFiles/PvPHero.dir/src/PvPHero.cpp.o CMakeFiles/PvPHero.dir/src/updating/Utils.cpp.o CMakeFiles/PvPHero.dir/src/updating/Downloader.cpp.o CMakeFiles/PvPHero.dir/src/updating/client/ClientUpdater.cpp.o CMakeFiles/PvPHero.dir/src/updating/launcher/LauncherUpdater.cpp.o CMakeFiles/PvPHero.dir/src/filesystem/InMemoryFileSystem.cpp.o CMakeFiles/PvPHero.dir/src/external/sago/platform_folders.cpp.o CMakeFiles/PvPHero.dir/src/external/zip/zip.c.o CMakeFiles/PvPHero.dir/src/Assets.cpp.o CMakeFiles/PvPHero.dir/src/main.cpp.o -o PvPHero.app/Contents/MacOS/PvPHero -L/Users/runner/work/pvphero-launcher/pvphero-launcher/build/../deps/AppCore/lib   -L/Users/runner/work/pvphero-launcher/pvphero-launcher/build/../deps/Ultralight/lib   -L/Users/runner/work/pvphero-launcher/pvphero-launcher/build/../deps/UltralightCore/lib   -L/Users/runner/work/pvphero-launcher/pvphero-launcher/build/../deps/WebCore/lib   -L/Users/runner/work/pvphero-launcher/pvphero-launcher/build/../deps/WebKitLibraries/lib -Wl,-rpath,@executable_path/  -lWebCore  -lUltralightCore  -lUltralight  -lAppCore  -lblend2d  -lWTF  -lJavaScriptCore  -lPAL  -lsqlite3  ../deps//WebKitLibraries/lib/libbrotlidec.a  ../deps//WebKitLibraries/lib/libbrotlienc.a  ../deps//WebKitLibraries/lib/libbrotlicommon.a  ../deps//WebKitLibraries/lib/libcrypto.a  ../deps//WebKitLibraries/lib/libcurl.a  ../deps//WebKitLibraries/lib/libfreetype.a  ../deps//WebKitLibraries/lib/libharfbuzz.a  ../deps//WebKitLibraries/lib/libharfbuzz-icu.a  ../deps//WebKitLibraries/lib/libicudata.a  ../deps//WebKitLibraries/lib/libicui18n.a  ../deps//WebKitLibraries/lib/libicuuc.a  ../deps//WebKitLibraries/lib/libjpeg.a  ../deps//WebKitLibraries/lib/libnghttp2_static.a  ../deps//WebKitLibraries/lib/libpng.a  ../deps//WebKitLibraries/lib/libpng16.a  ../deps//WebKitLibraries/lib/libssl.a  ../deps//WebKitLibraries/lib/libtls.a  ../deps//WebKitLibraries/lib/libxml2.a  ../deps//WebKitLibraries/lib/libxslt.a  ../deps//WebKitLibraries/lib/libz.a  -lpthread  -lbmalloc  -ldl  -lbz2  -framework  Accelerate  -framework  AppKit  -framework  SystemConfiguration  -framework  Security && :
clang: error: no such file or directory: '/SUBSYSTEM:CONSOLE'
ninja: build stopped: subcommand failed.
```

--- a/misc/output_test.py
+++ b/misc/output_test.py
@@ -152,14 +152,18 @@ class Output(unittest.TestCase):
         '',
     ))
 
-    def _test_expected_error(self, plan: str, flags: T.Optional[str], expected: str):
+    def _test_expected_error(self, plan: str, flags: T.Optional[str],expected: str,
+                             *args, exit_code: T.Optional[int]=None, **kwargs)->None:
         """Run Ninja with a given plan and flags, and verify its cooked output against an expected content.
+        All *args and **kwargs are passed to the `run` function
         """
         actual = ''
-        try:
-          actual = run(plan, flags, print_err_output=False)
-        except subprocess.CalledProcessError as err:
-          actual = err.cooked_output
+        kwargs['print_err_output'] = False
+        with self.assertRaises(subprocess.CalledProcessError) as cm:
+            run(plan, flags, *args,  **kwargs)
+        actual = cm.exception.cooked_output
+        if exit_code is not None:
+            self.assertEqual(cm.exception.returncode, exit_code)
         self.assertEqual(expected, actual)
 
     def test_issue_1418(self) -> None:
@@ -280,6 +284,75 @@ build zoo: touch bar
                 )
             except subprocess.CalledProcessError as err:
                 self.fail("non-zero exit code with: " + err.output)
+
+    def test_pr_2540(self)->None:
+        py = sys.executable
+        plan = f'''\
+rule CUSTOM_COMMAND
+  command = $COMMAND
+
+build 124: CUSTOM_COMMAND
+  COMMAND = {py} -c 'exit(124)'
+
+build 127: CUSTOM_COMMAND
+  COMMAND = {py} -c 'exit(127)'
+
+build 130: CUSTOM_COMMAND
+  COMMAND = {py} -c 'exit(130)'
+
+build 137: CUSTOM_COMMAND
+  COMMAND = {py} -c 'exit(137)'
+
+build success: CUSTOM_COMMAND
+  COMMAND = sleep 0.3; echo success
+'''
+        # Disable colors
+        env = default_env.copy()
+        env['TERM'] = 'dumb'
+        self._test_expected_error(
+            plan, '124',
+            f'''[1/1] {py} -c 'exit(124)'
+FAILED: [code=124] 124 \n{py} -c 'exit(124)'
+ninja: build stopped: subcommand failed.
+''',
+            exit_code=124, env=env,
+        )
+        self._test_expected_error(
+            plan, '127',
+            f'''[1/1] {py} -c 'exit(127)'
+FAILED: [code=127] 127 \n{py} -c 'exit(127)'
+ninja: build stopped: subcommand failed.
+''',
+            exit_code=127, env=env,
+        )
+        self._test_expected_error(
+            plan, '130',
+            'ninja: build stopped: interrupted by user.\n',
+            exit_code=130, env=env,
+        )
+        self._test_expected_error(
+            plan, '137',
+            f'''[1/1] {py} -c 'exit(137)'
+FAILED: [code=137] 137 \n{py} -c 'exit(137)'
+ninja: build stopped: subcommand failed.
+''',
+            exit_code=137, env=env,
+        )
+        self._test_expected_error(
+            plan, 'non-existent-target',
+            "ninja: error: unknown target 'non-existent-target'\n",
+            exit_code=1, env=env,
+        )
+        self._test_expected_error(
+            plan, '-j2 success 127',
+            f'''[1/2] {py} -c 'exit(127)'
+FAILED: [code=127] 127 \n{py} -c 'exit(127)'
+[2/2] sleep 0.3; echo success
+success
+ninja: build stopped: subcommand failed.
+''',
+            exit_code=127, env=env,
+        )
 
     def test_depfile_directory_creation(self) -> None:
         b = BuildDir('''\

--- a/src/build.cc
+++ b/src/build.cc
@@ -34,12 +34,12 @@
 #include "depfile_parser.h"
 #include "deps_log.h"
 #include "disk_interface.h"
+#include "exit_status.h"
 #include "explanations.h"
 #include "graph.h"
 #include "metrics.h"
 #include "state.h"
 #include "status.h"
-#include "subprocess.h"
 #include "util.h"
 
 using namespace std;
@@ -687,7 +687,7 @@ bool Builder::AlreadyUpToDate() const {
   return !plan_.more_to_do();
 }
 
-bool Builder::Build(string* err) {
+ExitStatus Builder::Build(string* err) {
   assert(!AlreadyUpToDate());
   plan_.PrepareQueue();
 
@@ -726,14 +726,14 @@ bool Builder::Build(string* err) {
         if (!StartEdge(edge, err)) {
           Cleanup();
           status_->BuildFinished();
-          return false;
+          return ExitFailure;
         }
 
         if (edge->is_phony()) {
           if (!plan_.EdgeFinished(edge, Plan::kEdgeSucceeded, err)) {
             Cleanup();
             status_->BuildFinished();
-            return false;
+            return ExitFailure;
           }
         } else {
           ++pending_commands;
@@ -760,14 +760,16 @@ bool Builder::Build(string* err) {
         Cleanup();
         status_->BuildFinished();
         *err = "interrupted by user";
-        return false;
+        return result.status;
       }
 
       --pending_commands;
-      if (!FinishCommand(&result, err)) {
+      bool command_finished = FinishCommand(&result, err);
+      SetFailureCode(result.status);
+      if (!command_finished) {
         Cleanup();
         status_->BuildFinished();
-        return false;
+        return result.status;
       }
 
       if (!result.success()) {
@@ -791,11 +793,11 @@ bool Builder::Build(string* err) {
     else
       *err = "stuck [this is a bug]";
 
-    return false;
+    return GetExitCode();
   }
 
   status_->BuildFinished();
-  return true;
+  return ExitSuccess;
 }
 
 bool Builder::StartEdge(Edge* edge, string* err) {
@@ -1035,4 +1037,11 @@ bool Builder::LoadDyndeps(Node* node, string* err) {
     return false;
 
   return true;
+}
+
+void Builder::SetFailureCode(ExitStatus code) {
+  // ExitSuccess should not overwrite any error
+  if (code != ExitSuccess) {
+    exit_code_ = code;
+  }
 }

--- a/src/build.cc
+++ b/src/build.cc
@@ -885,7 +885,7 @@ bool Builder::FinishCommand(CommandRunner::Result* result, string* err) {
   running_edges_.erase(it);
 
   status_->BuildEdgeFinished(edge, start_time_millis, end_time_millis,
-                             result->success(), result->output);
+                             result->status, result->output);
 
   // The rest of this function only applies to successful commands.
   if (!result->success()) {

--- a/src/build.h
+++ b/src/build.h
@@ -209,9 +209,9 @@ struct Builder {
   /// Returns true if the build targets are already up to date.
   bool AlreadyUpToDate() const;
 
-  /// Run the build.  Returns false on error.
+  /// Run the build.  Returns ExitStatus or the exit code of the last failed job.
   /// It is an error to call this function when AlreadyUpToDate() is true.
-  bool Build(std::string* err);
+  ExitStatus Build(std::string* err);
 
   bool StartEdge(Edge* edge, std::string* err);
 
@@ -233,6 +233,10 @@ struct Builder {
   std::unique_ptr<CommandRunner> command_runner_;
   Status* status_;
 
+  /// Returns ExitStatus or the exit code of the last failed job
+  /// (doesn't need to be an enum value of ExitStatus)
+  ExitStatus GetExitCode() const { return exit_code_; }
+
  private:
   bool ExtractDeps(CommandRunner::Result* result, const std::string& deps_type,
                    const std::string& deps_prefix,
@@ -252,6 +256,10 @@ struct Builder {
   std::unique_ptr<Explanations> explanations_;
 
   DependencyScan scan_;
+
+  /// Keep the global exit code for the build
+  ExitStatus exit_code_ = ExitSuccess;
+  void SetFailureCode(ExitStatus code);
 
   // Unimplemented copy ctor and operator= ensure we don't copy the auto_ptr.
   Builder(const Builder &other);        // DO NOT IMPLEMENT

--- a/src/build_test.cc
+++ b/src/build_test.cc
@@ -20,6 +20,7 @@
 
 #include "build_log.h"
 #include "deps_log.h"
+#include "exit_status.h"
 #include "graph.h"
 #include "status_printer.h"
 #include "test.h"
@@ -616,8 +617,8 @@ void BuildTest::RebuildTarget(const string& target, const char* manifest,
   command_runner_.commands_ran_.clear();
   builder.command_runner_.reset(&command_runner_);
   if (!builder.AlreadyUpToDate()) {
-    bool build_res = builder.Build(&err);
-    EXPECT_TRUE(build_res);
+    ExitStatus build_res = builder.Build(&err);
+    EXPECT_EQ(build_res, ExitSuccess);
   }
   builder.command_runner_.release();
 }
@@ -823,7 +824,7 @@ TEST_F(BuildTest, OneStep) {
   string err;
   EXPECT_TRUE(builder_.AddTarget("cat1", &err));
   ASSERT_EQ("", err);
-  EXPECT_TRUE(builder_.Build(&err));
+  EXPECT_EQ(builder_.Build(&err), ExitSuccess);
   ASSERT_EQ("", err);
 
   ASSERT_EQ(1u, command_runner_.commands_ran_.size());
@@ -837,7 +838,7 @@ TEST_F(BuildTest, OneStep2) {
   string err;
   EXPECT_TRUE(builder_.AddTarget("cat1", &err));
   ASSERT_EQ("", err);
-  EXPECT_TRUE(builder_.Build(&err));
+  EXPECT_EQ(builder_.Build(&err), ExitSuccess);
   EXPECT_EQ("", err);
 
   ASSERT_EQ(1u, command_runner_.commands_ran_.size());
@@ -848,7 +849,7 @@ TEST_F(BuildTest, TwoStep) {
   string err;
   EXPECT_TRUE(builder_.AddTarget("cat12", &err));
   ASSERT_EQ("", err);
-  EXPECT_TRUE(builder_.Build(&err));
+  EXPECT_EQ(builder_.Build(&err), ExitSuccess);
   EXPECT_EQ("", err);
   ASSERT_EQ(3u, command_runner_.commands_ran_.size());
   // Depending on how the pointers work out, we could've ran
@@ -868,7 +869,7 @@ TEST_F(BuildTest, TwoStep) {
   state_.Reset();
   EXPECT_TRUE(builder_.AddTarget("cat12", &err));
   ASSERT_EQ("", err);
-  EXPECT_TRUE(builder_.Build(&err));
+  EXPECT_EQ(builder_.Build(&err), ExitSuccess);
   ASSERT_EQ("", err);
   ASSERT_EQ(5u, command_runner_.commands_ran_.size());
   EXPECT_EQ("cat in1 in2 > cat2", command_runner_.commands_ran_[3]);
@@ -886,7 +887,7 @@ TEST_F(BuildTest, TwoOutputs) {
   string err;
   EXPECT_TRUE(builder_.AddTarget("out1", &err));
   ASSERT_EQ("", err);
-  EXPECT_TRUE(builder_.Build(&err));
+  EXPECT_EQ(builder_.Build(&err), ExitSuccess);
   EXPECT_EQ("", err);
   ASSERT_EQ(1u, command_runner_.commands_ran_.size());
   EXPECT_EQ("touch out1 out2", command_runner_.commands_ran_[0]);
@@ -902,7 +903,7 @@ TEST_F(BuildTest, ImplicitOutput) {
   string err;
   EXPECT_TRUE(builder_.AddTarget("out.imp", &err));
   ASSERT_EQ("", err);
-  EXPECT_TRUE(builder_.Build(&err));
+  EXPECT_EQ(builder_.Build(&err), ExitSuccess);
   EXPECT_EQ("", err);
   ASSERT_EQ(1u, command_runner_.commands_ran_.size());
   EXPECT_EQ("touch out out.imp", command_runner_.commands_ran_[0]);
@@ -924,7 +925,7 @@ TEST_F(BuildTest, MultiOutIn) {
   string err;
   EXPECT_TRUE(builder_.AddTarget("out", &err));
   ASSERT_EQ("", err);
-  EXPECT_TRUE(builder_.Build(&err));
+  EXPECT_EQ(builder_.Build(&err), ExitSuccess);
   EXPECT_EQ("", err);
 }
 
@@ -940,7 +941,7 @@ TEST_F(BuildTest, Chain) {
   string err;
   EXPECT_TRUE(builder_.AddTarget("c5", &err));
   ASSERT_EQ("", err);
-  EXPECT_TRUE(builder_.Build(&err));
+  EXPECT_EQ(builder_.Build(&err), ExitSuccess);
   EXPECT_EQ("", err);
   ASSERT_EQ(4u, command_runner_.commands_ran_.size());
 
@@ -960,7 +961,7 @@ TEST_F(BuildTest, Chain) {
   EXPECT_TRUE(builder_.AddTarget("c5", &err));
   ASSERT_EQ("", err);
   EXPECT_FALSE(builder_.AlreadyUpToDate());
-  EXPECT_TRUE(builder_.Build(&err));
+  EXPECT_EQ(builder_.Build(&err), ExitSuccess);
   ASSERT_EQ(2u, command_runner_.commands_ran_.size());  // 3->4, 4->5
 }
 
@@ -1001,7 +1002,7 @@ TEST_F(BuildTest, MakeDirs) {
   EXPECT_TRUE(builder_.AddTarget("subdir/dir2/file", &err));
 
   EXPECT_EQ("", err);
-  EXPECT_TRUE(builder_.Build(&err));
+  EXPECT_EQ(builder_.Build(&err), ExitSuccess);
   ASSERT_EQ("", err);
   ASSERT_EQ(2u, fs_.directories_made_.size());
   EXPECT_EQ("subdir", fs_.directories_made_[0]);
@@ -1086,7 +1087,7 @@ TEST_F(BuildTest, EncounterReadyTwice) {
   EXPECT_TRUE(builder_.AddTarget("a", &err));
   ASSERT_EQ("", err);
 
-  EXPECT_TRUE(builder_.Build(&err));
+  EXPECT_EQ(builder_.Build(&err), ExitSuccess);
   ASSERT_EQ("", err);
   ASSERT_EQ(2u, command_runner_.commands_ran_.size());
 }
@@ -1119,7 +1120,7 @@ TEST_F(BuildTest, OrderOnlyDeps) {
   ASSERT_EQ("cc foo.c", edge->EvaluateCommand());
 
   // explicit dep dirty, expect a rebuild.
-  EXPECT_TRUE(builder_.Build(&err));
+  EXPECT_EQ(builder_.Build(&err), ExitSuccess);
   ASSERT_EQ("", err);
   ASSERT_EQ(1u, command_runner_.commands_ran_.size());
 
@@ -1134,7 +1135,7 @@ TEST_F(BuildTest, OrderOnlyDeps) {
   command_runner_.commands_ran_.clear();
   state_.Reset();
   EXPECT_TRUE(builder_.AddTarget("foo.o", &err));
-  EXPECT_TRUE(builder_.Build(&err));
+  EXPECT_EQ(builder_.Build(&err), ExitSuccess);
   ASSERT_EQ("", err);
   ASSERT_EQ(1u, command_runner_.commands_ran_.size());
 
@@ -1156,7 +1157,7 @@ TEST_F(BuildTest, OrderOnlyDeps) {
   command_runner_.commands_ran_.clear();
   state_.Reset();
   EXPECT_TRUE(builder_.AddTarget("foo.o", &err));
-  EXPECT_TRUE(builder_.Build(&err));
+  EXPECT_EQ(builder_.Build(&err), ExitSuccess);
   ASSERT_EQ("", err);
   ASSERT_EQ(1u, command_runner_.commands_ran_.size());
 }
@@ -1174,7 +1175,7 @@ TEST_F(BuildTest, RebuildOrderOnlyDeps) {
 
   // foo.o and order-only dep dirty, build both.
   EXPECT_TRUE(builder_.AddTarget("foo.o", &err));
-  EXPECT_TRUE(builder_.Build(&err));
+  EXPECT_EQ(builder_.Build(&err), ExitSuccess);
   ASSERT_EQ("", err);
   ASSERT_EQ(2u, command_runner_.commands_ran_.size());
 
@@ -1190,7 +1191,7 @@ TEST_F(BuildTest, RebuildOrderOnlyDeps) {
   command_runner_.commands_ran_.clear();
   state_.Reset();
   EXPECT_TRUE(builder_.AddTarget("foo.o", &err));
-  EXPECT_TRUE(builder_.Build(&err));
+  EXPECT_EQ(builder_.Build(&err), ExitSuccess);
   ASSERT_EQ("", err);
   ASSERT_EQ(1u, command_runner_.commands_ran_.size());
   ASSERT_EQ("cc oo.h.in", command_runner_.commands_ran_[0]);
@@ -1202,7 +1203,7 @@ TEST_F(BuildTest, RebuildOrderOnlyDeps) {
   command_runner_.commands_ran_.clear();
   state_.Reset();
   EXPECT_TRUE(builder_.AddTarget("foo.o", &err));
-  EXPECT_TRUE(builder_.Build(&err));
+  EXPECT_EQ(builder_.Build(&err), ExitSuccess);
   ASSERT_EQ("", err);
   ASSERT_EQ(1u, command_runner_.commands_ran_.size());
   ASSERT_EQ("cc oo.h.in", command_runner_.commands_ran_[0]);
@@ -1251,7 +1252,7 @@ TEST_F(BuildTest, Phony) {
 
   // Only one command to run, because phony runs no command.
   EXPECT_FALSE(builder_.AlreadyUpToDate());
-  EXPECT_TRUE(builder_.Build(&err));
+  EXPECT_EQ(builder_.Build(&err), ExitSuccess);
   ASSERT_EQ("", err);
   ASSERT_EQ(1u, command_runner_.commands_ran_.size());
 }
@@ -1345,7 +1346,7 @@ void TestPhonyUseCase(BuildTest* t, int i) {
   ASSERT_EQ("", err);
   EXPECT_TRUE(builder_.AddTarget("test6", &err));
   ASSERT_EQ("", err);
-  EXPECT_TRUE(builder_.Build(&err));
+  EXPECT_EQ(builder_.Build(&err), ExitSuccess);
   ASSERT_EQ("", err);
 
   string ci;
@@ -1364,7 +1365,7 @@ void TestPhonyUseCase(BuildTest* t, int i) {
     EXPECT_TRUE(builder_.AddTarget("test" + ci, &err));
     ASSERT_EQ("", err);
     if (!builder_.AlreadyUpToDate()) {
-      EXPECT_TRUE(builder_.Build(&err));
+      EXPECT_EQ(builder_.Build(&err), ExitSuccess);
     }
     ASSERT_EQ("", err);
 
@@ -1379,7 +1380,7 @@ void TestPhonyUseCase(BuildTest* t, int i) {
     // Second build, expect testN edge to be rebuilt
     // and phonyN node's mtime to be updated.
     EXPECT_FALSE(builder_.AlreadyUpToDate());
-    EXPECT_TRUE(builder_.Build(&err));
+    EXPECT_EQ(builder_.Build(&err), ExitSuccess);
     ASSERT_EQ("", err);
     ASSERT_EQ(1u, command_runner_.commands_ran_.size());
     EXPECT_EQ(string("touch test") + ci, command_runner_.commands_ran_[0]);
@@ -1405,7 +1406,7 @@ void TestPhonyUseCase(BuildTest* t, int i) {
     EXPECT_TRUE(builder_.AddTarget("test" + ci, &err));
     ASSERT_EQ("", err);
     EXPECT_FALSE(builder_.AlreadyUpToDate());
-    EXPECT_TRUE(builder_.Build(&err));
+    EXPECT_EQ(builder_.Build(&err), ExitSuccess);
     ASSERT_EQ("", err);
     ASSERT_EQ(1u, command_runner_.commands_ran_.size());
     EXPECT_EQ("touch test" + ci, command_runner_.commands_ran_[0]);
@@ -1415,7 +1416,7 @@ void TestPhonyUseCase(BuildTest* t, int i) {
     EXPECT_TRUE(builder_.AddTarget("test" + ci, &err));
     ASSERT_EQ("", err);
     EXPECT_FALSE(builder_.AlreadyUpToDate());
-    EXPECT_TRUE(builder_.Build(&err));
+    EXPECT_EQ(builder_.Build(&err), ExitSuccess);
     ASSERT_EQ("", err);
     ASSERT_EQ(1u, command_runner_.commands_ran_.size());
     EXPECT_EQ("touch test" + ci, command_runner_.commands_ran_[0]);
@@ -1439,7 +1440,7 @@ TEST_F(BuildTest, Fail) {
   EXPECT_TRUE(builder_.AddTarget("out1", &err));
   ASSERT_EQ("", err);
 
-  EXPECT_FALSE(builder_.Build(&err));
+  EXPECT_EQ(builder_.Build(&err), ExitFailure);
   ASSERT_EQ(1u, command_runner_.commands_ran_.size());
   ASSERT_EQ("subcommand failed", err);
 }
@@ -1460,7 +1461,7 @@ TEST_F(BuildTest, SwallowFailures) {
   EXPECT_TRUE(builder_.AddTarget("all", &err));
   ASSERT_EQ("", err);
 
-  EXPECT_FALSE(builder_.Build(&err));
+  EXPECT_EQ(builder_.Build(&err), ExitFailure);
   ASSERT_EQ(3u, command_runner_.commands_ran_.size());
   ASSERT_EQ("subcommands failed", err);
 }
@@ -1481,7 +1482,7 @@ TEST_F(BuildTest, SwallowFailuresLimit) {
   EXPECT_TRUE(builder_.AddTarget("final", &err));
   ASSERT_EQ("", err);
 
-  EXPECT_FALSE(builder_.Build(&err));
+  EXPECT_EQ(builder_.Build(&err), ExitFailure);
   ASSERT_EQ(3u, command_runner_.commands_ran_.size());
   ASSERT_EQ("cannot make progress due to previous errors", err);
 }
@@ -1505,7 +1506,7 @@ TEST_F(BuildTest, SwallowFailuresPool) {
   EXPECT_TRUE(builder_.AddTarget("final", &err));
   ASSERT_EQ("", err);
 
-  EXPECT_FALSE(builder_.Build(&err));
+  EXPECT_EQ(builder_.Build(&err), ExitFailure);
   ASSERT_EQ(3u, command_runner_.commands_ran_.size());
   ASSERT_EQ("cannot make progress due to previous errors", err);
 }
@@ -1579,7 +1580,7 @@ TEST_F(BuildWithLogTest, ImplicitGeneratedOutOfDate2) {
   EXPECT_TRUE(builder_.AddTarget("out.imp", &err));
   EXPECT_FALSE(builder_.AlreadyUpToDate());
 
-  EXPECT_TRUE(builder_.Build(&err));
+  EXPECT_EQ(builder_.Build(&err), ExitSuccess);
   EXPECT_TRUE(builder_.AlreadyUpToDate());
 
   command_runner_.commands_ran_.clear();
@@ -1602,7 +1603,7 @@ TEST_F(BuildWithLogTest, ImplicitGeneratedOutOfDate2) {
   EXPECT_TRUE(builder_.AddTarget("out.imp", &err));
   EXPECT_FALSE(builder_.AlreadyUpToDate());
 
-  EXPECT_TRUE(builder_.Build(&err));
+  EXPECT_EQ(builder_.Build(&err), ExitSuccess);
   EXPECT_TRUE(builder_.AlreadyUpToDate());
 
   command_runner_.commands_ran_.clear();
@@ -1636,7 +1637,7 @@ TEST_F(BuildWithLogTest, NotInLogButOnDisk) {
   state_.Reset();
 
   EXPECT_TRUE(builder_.AddTarget("out1", &err));
-  EXPECT_TRUE(builder_.Build(&err));
+  EXPECT_EQ(builder_.Build(&err), ExitSuccess);
   EXPECT_TRUE(builder_.AlreadyUpToDate());
 }
 
@@ -1652,7 +1653,7 @@ TEST_F(BuildWithLogTest, RebuildAfterFailure) {
 
   // Run once successfully to get out1 in the log
   EXPECT_TRUE(builder_.AddTarget("out1", &err));
-  EXPECT_TRUE(builder_.Build(&err));
+  EXPECT_EQ(builder_.Build(&err), ExitSuccess);
   EXPECT_EQ("", err);
   EXPECT_EQ(1u, command_runner_.commands_ran_.size());
 
@@ -1666,7 +1667,7 @@ TEST_F(BuildWithLogTest, RebuildAfterFailure) {
 
   // Run again with a failure that updates the output file timestamp
   EXPECT_TRUE(builder_.AddTarget("out1", &err));
-  EXPECT_FALSE(builder_.Build(&err));
+  EXPECT_EQ(builder_.Build(&err), ExitFailure);
   EXPECT_EQ("subcommand failed", err);
   EXPECT_EQ(1u, command_runner_.commands_ran_.size());
 
@@ -1680,7 +1681,7 @@ TEST_F(BuildWithLogTest, RebuildAfterFailure) {
   // Run again, should rerun even though the output file is up to date on disk
   EXPECT_TRUE(builder_.AddTarget("out1", &err));
   EXPECT_FALSE(builder_.AlreadyUpToDate());
-  EXPECT_TRUE(builder_.Build(&err));
+  EXPECT_EQ(builder_.Build(&err), ExitSuccess);
   EXPECT_EQ(1u, command_runner_.commands_ran_.size());
   EXPECT_EQ("", err);
 }
@@ -1698,7 +1699,7 @@ TEST_F(BuildWithLogTest, RebuildWithNoInputs) {
 
   EXPECT_TRUE(builder_.AddTarget("out1", &err));
   EXPECT_TRUE(builder_.AddTarget("out2", &err));
-  EXPECT_TRUE(builder_.Build(&err));
+  EXPECT_EQ(builder_.Build(&err), ExitSuccess);
   EXPECT_EQ("", err);
   EXPECT_EQ(2u, command_runner_.commands_ran_.size());
 
@@ -1711,7 +1712,7 @@ TEST_F(BuildWithLogTest, RebuildWithNoInputs) {
 
   EXPECT_TRUE(builder_.AddTarget("out1", &err));
   EXPECT_TRUE(builder_.AddTarget("out2", &err));
-  EXPECT_TRUE(builder_.Build(&err));
+  EXPECT_EQ(builder_.Build(&err), ExitSuccess);
   EXPECT_EQ("", err);
   EXPECT_EQ(1u, command_runner_.commands_ran_.size());
 }
@@ -1742,7 +1743,7 @@ TEST_F(BuildWithLogTest, RestatTest) {
   string err;
   EXPECT_TRUE(builder_.AddTarget("out3", &err));
   ASSERT_EQ("", err);
-  EXPECT_TRUE(builder_.Build(&err));
+  EXPECT_EQ(builder_.Build(&err), ExitSuccess);
   ASSERT_EQ("", err);
   EXPECT_EQ(3u, command_runner_.commands_ran_.size());
   EXPECT_EQ(3u, builder_.plan_.command_edge_count());
@@ -1756,7 +1757,7 @@ TEST_F(BuildWithLogTest, RestatTest) {
   // touch out2, we should cancel the build of out3.
   EXPECT_TRUE(builder_.AddTarget("out3", &err));
   ASSERT_EQ("", err);
-  EXPECT_TRUE(builder_.Build(&err));
+  EXPECT_EQ(builder_.Build(&err), ExitSuccess);
   ASSERT_EQ(2u, command_runner_.commands_ran_.size());
 
   // If we run again, it should be a no-op, because the build log has recorded
@@ -1777,7 +1778,7 @@ TEST_F(BuildWithLogTest, RestatTest) {
   state_.Reset();
   EXPECT_TRUE(builder_.AddTarget("out3", &err));
   ASSERT_EQ("", err);
-  EXPECT_TRUE(builder_.Build(&err));
+  EXPECT_EQ(builder_.Build(&err), ExitSuccess);
   ASSERT_EQ(2u, command_runner_.commands_ran_.size());
 }
 
@@ -1804,7 +1805,7 @@ TEST_F(BuildWithLogTest, RestatMissingFile) {
   string err;
   EXPECT_TRUE(builder_.AddTarget("out2", &err));
   ASSERT_EQ("", err);
-  EXPECT_TRUE(builder_.Build(&err));
+  EXPECT_EQ(builder_.Build(&err), ExitSuccess);
   ASSERT_EQ("", err);
   command_runner_.commands_ran_.clear();
   state_.Reset();
@@ -1818,7 +1819,7 @@ TEST_F(BuildWithLogTest, RestatMissingFile) {
   // we shouldn't run the dependent build.
   EXPECT_TRUE(builder_.AddTarget("out2", &err));
   ASSERT_EQ("", err);
-  EXPECT_TRUE(builder_.Build(&err));
+  EXPECT_EQ(builder_.Build(&err), ExitSuccess);
   ASSERT_EQ(1u, command_runner_.commands_ran_.size());
 }
 
@@ -1840,7 +1841,7 @@ TEST_F(BuildWithLogTest, RestatSingleDependentOutputDirty) {
   string err;
   EXPECT_TRUE(builder_.AddTarget("out4", &err));
   ASSERT_EQ("", err);
-  EXPECT_TRUE(builder_.Build(&err));
+  EXPECT_EQ(builder_.Build(&err), ExitSuccess);
   ASSERT_EQ("", err);
   ASSERT_EQ(3u, command_runner_.commands_ran_.size());
 
@@ -1857,7 +1858,7 @@ TEST_F(BuildWithLogTest, RestatSingleDependentOutputDirty) {
   state_.Reset();
   EXPECT_TRUE(builder_.AddTarget("out4", &err));
   ASSERT_EQ("", err);
-  EXPECT_TRUE(builder_.Build(&err));
+  EXPECT_EQ(builder_.Build(&err), ExitSuccess);
   ASSERT_EQ("", err);
   ASSERT_EQ(3u, command_runner_.commands_ran_.size());
 }
@@ -1889,7 +1890,7 @@ TEST_F(BuildWithLogTest, RestatMissingInput) {
   string err;
   EXPECT_TRUE(builder_.AddTarget("out2", &err));
   ASSERT_EQ("", err);
-  EXPECT_TRUE(builder_.Build(&err));
+  EXPECT_EQ(builder_.Build(&err), ExitSuccess);
   ASSERT_EQ(2u, command_runner_.commands_ran_.size());
 
   // See that an entry in the logfile is created, capturing
@@ -1907,7 +1908,7 @@ TEST_F(BuildWithLogTest, RestatMissingInput) {
   state_.Reset();
   EXPECT_TRUE(builder_.AddTarget("out2", &err));
   ASSERT_EQ("", err);
-  EXPECT_TRUE(builder_.Build(&err));
+  EXPECT_EQ(builder_.Build(&err), ExitSuccess);
   ASSERT_EQ(1u, command_runner_.commands_ran_.size());
 
   // Check that the logfile entry remains correctly set
@@ -1933,7 +1934,7 @@ TEST_F(BuildWithLogTest, RestatInputChangesDueToRule) {
   string err;
   EXPECT_TRUE(builder_.AddTarget("out1", &err));
   ASSERT_EQ("", err);
-  EXPECT_TRUE(builder_.Build(&err));
+  EXPECT_EQ(builder_.Build(&err), ExitSuccess);
   ASSERT_EQ("", err);
   EXPECT_EQ(2u, command_runner_.commands_ran_.size());
   EXPECT_EQ(2u, builder_.plan_.command_edge_count());
@@ -1956,7 +1957,7 @@ TEST_F(BuildWithLogTest, RestatInputChangesDueToRule) {
   EXPECT_TRUE(builder_.AddTarget("out1", &err));
   ASSERT_EQ("", err);
   EXPECT_TRUE(!state_.GetNode("out1", 0)->dirty());
-  EXPECT_TRUE(builder_.Build(&err));
+  EXPECT_EQ(builder_.Build(&err), ExitSuccess);
   ASSERT_EQ("", err);
   EXPECT_EQ(1u, command_runner_.commands_ran_.size());
   EXPECT_EQ(1u, builder_.plan_.command_edge_count());
@@ -1977,7 +1978,7 @@ TEST_F(BuildWithLogTest, GeneratedPlainDepfileMtime) {
   EXPECT_TRUE(builder_.AddTarget("out", &err));
   EXPECT_FALSE(builder_.AlreadyUpToDate());
 
-  EXPECT_TRUE(builder_.Build(&err));
+  EXPECT_EQ(builder_.Build(&err), ExitSuccess);
   EXPECT_TRUE(builder_.AlreadyUpToDate());
 
   command_runner_.commands_ran_.clear();
@@ -2020,7 +2021,7 @@ TEST_F(BuildDryRun, AllCommandsShown) {
   string err;
   EXPECT_TRUE(builder_.AddTarget("out3", &err));
   ASSERT_EQ("", err);
-  EXPECT_TRUE(builder_.Build(&err));
+  EXPECT_EQ(builder_.Build(&err), ExitSuccess);
   ASSERT_EQ(3u, command_runner_.commands_ran_.size());
 }
 
@@ -2063,7 +2064,7 @@ TEST_F(BuildTest, RspFileSuccess)
   size_t files_created = fs_.files_created_.size();
   size_t files_removed = fs_.files_removed_.size();
 
-  EXPECT_TRUE(builder_.Build(&err));
+  EXPECT_EQ(builder_.Build(&err), ExitSuccess);
   ASSERT_EQ(3u, command_runner_.commands_ran_.size());
 
   // The RSP files and temp file to acquire output mtimes were created
@@ -2100,7 +2101,7 @@ TEST_F(BuildTest, RspFileFailure) {
   size_t files_created = fs_.files_created_.size();
   size_t files_removed = fs_.files_removed_.size();
 
-  EXPECT_FALSE(builder_.Build(&err));
+  EXPECT_EQ(builder_.Build(&err), ExitFailure);
   ASSERT_EQ("subcommand failed", err);
   ASSERT_EQ(1u, command_runner_.commands_ran_.size());
 
@@ -2138,7 +2139,7 @@ TEST_F(BuildWithLogTest, RspFileCmdLineChange) {
   ASSERT_EQ("", err);
 
   // 1. Build for the 1st time (-> populate log)
-  EXPECT_TRUE(builder_.Build(&err));
+  EXPECT_EQ(builder_.Build(&err), ExitSuccess);
   ASSERT_EQ(1u, command_runner_.commands_ran_.size());
 
   // 2. Build again (no change)
@@ -2161,7 +2162,7 @@ TEST_F(BuildWithLogTest, RspFileCmdLineChange) {
   state_.Reset();
   EXPECT_TRUE(builder_.AddTarget("out", &err));
   EXPECT_EQ("", err);
-  EXPECT_TRUE(builder_.Build(&err));
+  EXPECT_EQ(builder_.Build(&err), ExitSuccess);
   EXPECT_EQ(1u, command_runner_.commands_ran_.size());
 }
 
@@ -2184,7 +2185,7 @@ TEST_F(BuildTest, InterruptCleanup) {
   string err;
   EXPECT_TRUE(builder_.AddTarget("out1", &err));
   EXPECT_EQ("", err);
-  EXPECT_FALSE(builder_.Build(&err));
+  EXPECT_EQ(builder_.Build(&err), ExitInterrupted);
   EXPECT_EQ("interrupted by user", err);
   builder_.Cleanup();
   EXPECT_GT(fs_.Stat("out1", &err), 0);
@@ -2193,7 +2194,7 @@ TEST_F(BuildTest, InterruptCleanup) {
   // A touched output of an interrupted command should be deleted.
   EXPECT_TRUE(builder_.AddTarget("out2", &err));
   EXPECT_EQ("", err);
-  EXPECT_FALSE(builder_.Build(&err));
+  EXPECT_EQ(builder_.Build(&err), ExitInterrupted);
   EXPECT_EQ("interrupted by user", err);
   builder_.Cleanup();
   EXPECT_EQ(0, fs_.Stat("out2", &err));
@@ -2235,7 +2236,7 @@ TEST_F(BuildTest, PhonyWithNoInputs) {
   state_.Reset();
   EXPECT_TRUE(builder_.AddTarget("out2", &err));
   ASSERT_EQ("", err);
-  EXPECT_TRUE(builder_.Build(&err));
+  EXPECT_EQ(builder_.Build(&err), ExitSuccess);
   EXPECT_EQ("", err);
   ASSERT_EQ(1u, command_runner_.commands_ran_.size());
 }
@@ -2253,7 +2254,7 @@ TEST_F(BuildTest, DepsGccWithEmptyDepfileErrorsOut) {
   ASSERT_EQ("", err);
   EXPECT_FALSE(builder_.AlreadyUpToDate());
 
-  EXPECT_FALSE(builder_.Build(&err));
+  EXPECT_EQ(builder_.Build(&err), ExitFailure);
   ASSERT_EQ("subcommand failed", err);
   ASSERT_EQ(1u, command_runner_.commands_ran_.size());
 }
@@ -2301,7 +2302,7 @@ TEST_F(BuildTest, FailedDepsParse) {
   // path to the left of the colon.
   fs_.Create("in1.d", "AAA BBB");
 
-  EXPECT_FALSE(builder_.Build(&err));
+  EXPECT_EQ(builder_.Build(&err), ExitFailure);
   EXPECT_EQ("subcommand failed", err);
 }
 
@@ -2341,7 +2342,7 @@ TEST_F(BuildWithQueryDepsLogTest, TwoOutputsDepFileMSVC) {
   std::string err;
   EXPECT_TRUE(builder_.AddTarget("out1", &err));
   ASSERT_EQ("", err);
-  EXPECT_TRUE(builder_.Build(&err));
+  EXPECT_EQ(builder_.Build(&err), ExitSuccess);
   EXPECT_EQ("", err);
   ASSERT_EQ(1u, command_runner_.commands_ran_.size());
   EXPECT_EQ("echo 'using in1' && for file in out1 out2; do cp in1 $file; done", command_runner_.commands_ran_[0]);
@@ -2370,7 +2371,7 @@ TEST_F(BuildWithQueryDepsLogTest, TwoOutputsDepFileGCCOneLine) {
   EXPECT_TRUE(builder_.AddTarget("out1", &err));
   ASSERT_EQ("", err);
   fs_.Create("in.d", "out1 out2: in1 in2");
-  EXPECT_TRUE(builder_.Build(&err));
+  EXPECT_EQ(builder_.Build(&err), ExitSuccess);
   EXPECT_EQ("", err);
   ASSERT_EQ(1u, command_runner_.commands_ran_.size());
   EXPECT_EQ("echo 'out1 out2: in1 in2' > in.d && for file in out1 out2; do cp in1 $file; done", command_runner_.commands_ran_[0]);
@@ -2401,7 +2402,7 @@ TEST_F(BuildWithQueryDepsLogTest, TwoOutputsDepFileGCCMultiLineInput) {
   EXPECT_TRUE(builder_.AddTarget("out1", &err));
   ASSERT_EQ("", err);
   fs_.Create("in.d", "out1 out2: in1\nout1 out2: in2");
-  EXPECT_TRUE(builder_.Build(&err));
+  EXPECT_EQ(builder_.Build(&err), ExitSuccess);
   EXPECT_EQ("", err);
   ASSERT_EQ(1u, command_runner_.commands_ran_.size());
   EXPECT_EQ("echo 'out1 out2: in1\\nout1 out2: in2' > in.d && for file in out1 out2; do cp in1 $file; done", command_runner_.commands_ran_[0]);
@@ -2432,7 +2433,7 @@ TEST_F(BuildWithQueryDepsLogTest, TwoOutputsDepFileGCCMultiLineOutput) {
   EXPECT_TRUE(builder_.AddTarget("out1", &err));
   ASSERT_EQ("", err);
   fs_.Create("in.d", "out1: in1 in2\nout2: in1 in2");
-  EXPECT_TRUE(builder_.Build(&err));
+  EXPECT_EQ(builder_.Build(&err), ExitSuccess);
   EXPECT_EQ("", err);
   ASSERT_EQ(1u, command_runner_.commands_ran_.size());
   EXPECT_EQ("echo 'out1: in1 in2\\nout2: in1 in2' > in.d && for file in out1 out2; do cp in1 $file; done", command_runner_.commands_ran_[0]);
@@ -2463,7 +2464,7 @@ TEST_F(BuildWithQueryDepsLogTest, TwoOutputsDepFileGCCOnlyMainOutput) {
   EXPECT_TRUE(builder_.AddTarget("out1", &err));
   ASSERT_EQ("", err);
   fs_.Create("in.d", "out1: in1 in2");
-  EXPECT_TRUE(builder_.Build(&err));
+  EXPECT_EQ(builder_.Build(&err), ExitSuccess);
   EXPECT_EQ("", err);
   ASSERT_EQ(1u, command_runner_.commands_ran_.size());
   EXPECT_EQ("echo 'out1: in1 in2' > in.d && for file in out1 out2; do cp in1 $file; done", command_runner_.commands_ran_[0]);
@@ -2496,7 +2497,7 @@ TEST_F(BuildWithQueryDepsLogTest, TwoOutputsDepFileGCCOnlySecondaryOutput) {
   EXPECT_TRUE(builder_.AddTarget("out1", &err));
   ASSERT_EQ("", err);
   fs_.Create("in.d", "out2: in1 in2");
-  EXPECT_TRUE(builder_.Build(&err));
+  EXPECT_EQ(builder_.Build(&err), ExitSuccess);
   EXPECT_EQ("", err);
   ASSERT_EQ(1u, command_runner_.commands_ran_.size());
   EXPECT_EQ("echo 'out2: in1 in2' > in.d && for file in out1 out2; do cp in1 $file; done", command_runner_.commands_ran_[0]);
@@ -2564,7 +2565,7 @@ TEST_F(BuildWithDepsLogTest, Straightforward) {
     EXPECT_TRUE(builder.AddTarget("out", &err));
     ASSERT_EQ("", err);
     fs_.Create("in1.d", "out: in2");
-    EXPECT_TRUE(builder.Build(&err));
+    EXPECT_EQ(builder.Build(&err), ExitSuccess);
     EXPECT_EQ("", err);
 
     // The deps file should have been removed.
@@ -2594,7 +2595,7 @@ TEST_F(BuildWithDepsLogTest, Straightforward) {
     command_runner_.commands_ran_.clear();
     EXPECT_TRUE(builder.AddTarget("out", &err));
     ASSERT_EQ("", err);
-    EXPECT_TRUE(builder.Build(&err));
+    EXPECT_EQ(builder.Build(&err), ExitSuccess);
     EXPECT_EQ("", err);
 
     // We should have rebuilt the output due to in2 being
@@ -2634,7 +2635,7 @@ TEST_F(BuildWithDepsLogTest, ObsoleteDeps) {
     builder.command_runner_.reset(&command_runner_);
     EXPECT_TRUE(builder.AddTarget("out", &err));
     ASSERT_EQ("", err);
-    EXPECT_TRUE(builder.Build(&err));
+    EXPECT_EQ(builder.Build(&err), ExitSuccess);
     EXPECT_EQ("", err);
 
     deps_log.Close();
@@ -2668,7 +2669,7 @@ TEST_F(BuildWithDepsLogTest, ObsoleteDeps) {
     // Recreate the deps file here because the build expects them to exist.
     fs_.Create("in1.d", "out: ");
 
-    EXPECT_TRUE(builder.Build(&err));
+    EXPECT_EQ(builder.Build(&err), ExitSuccess);
     EXPECT_EQ("", err);
 
     // We should have rebuilt the output due to the deps being
@@ -2702,7 +2703,7 @@ TEST_F(BuildWithDepsLogTest, DepsIgnoredInDryRun) {
   string err;
   EXPECT_TRUE(builder.AddTarget("out", &err));
   ASSERT_EQ("", err);
-  EXPECT_TRUE(builder.Build(&err));
+  EXPECT_EQ(builder.Build(&err), ExitSuccess);
   ASSERT_EQ(1u, command_runner_.commands_ran_.size());
 
   builder.command_runner_.release();
@@ -2737,7 +2738,7 @@ TEST_F(BuildWithDepsLogTest, TestInputMtimeRaceCondition) {
     // Run the build, out gets built, dep file is created
     EXPECT_TRUE(builder.AddTarget("out", &err));
     ASSERT_EQ("", err);
-    EXPECT_TRUE(builder.Build(&err));
+    EXPECT_EQ(builder.Build(&err), ExitSuccess);
     ASSERT_EQ(1u, command_runner_.commands_ran_.size());
 
     // See that an entry in the logfile is created. the input_mtime is 1 since that was
@@ -2761,7 +2762,7 @@ TEST_F(BuildWithDepsLogTest, TestInputMtimeRaceCondition) {
     state.Reset();
     EXPECT_TRUE(builder.AddTarget("out", &err));
     ASSERT_EQ("", err);
-    EXPECT_TRUE(builder.Build(&err));
+    EXPECT_EQ(builder.Build(&err), ExitSuccess);
     ASSERT_EQ(1u, command_runner_.commands_ran_.size());
 
     // Check that the logfile entry is still correct
@@ -2817,7 +2818,7 @@ TEST_F(BuildWithDepsLogTest, TestInputMtimeRaceConditionWithDepFile) {
     // Run the build, out gets built, dep file is created
     EXPECT_TRUE(builder.AddTarget("out", &err));
     ASSERT_EQ("", err);
-    EXPECT_TRUE(builder.Build(&err));
+    EXPECT_EQ(builder.Build(&err), ExitSuccess);
     ASSERT_EQ(1u, command_runner_.commands_ran_.size());
 
     // See that an entry in the logfile is created. the mtime is 1 due to the command
@@ -2839,7 +2840,7 @@ TEST_F(BuildWithDepsLogTest, TestInputMtimeRaceConditionWithDepFile) {
     state.Reset();
     EXPECT_TRUE(builder.AddTarget("out", &err));
     ASSERT_EQ("", err);
-    EXPECT_TRUE(builder.Build(&err));
+    EXPECT_EQ(builder.Build(&err), ExitSuccess);
     ASSERT_EQ(1u, command_runner_.commands_ran_.size());
 
     builder.command_runner_.release();
@@ -2874,7 +2875,7 @@ TEST_F(BuildWithDepsLogTest, TestInputMtimeRaceConditionWithDepFile) {
     state.Reset();
     EXPECT_TRUE(builder.AddTarget("out", &err));
     ASSERT_EQ("", err);
-    EXPECT_TRUE(builder.Build(&err));
+    EXPECT_EQ(builder.Build(&err), ExitSuccess);
     ASSERT_EQ(1u, command_runner_.commands_ran_.size());
 
     builder.command_runner_.release();
@@ -2890,7 +2891,7 @@ TEST_F(BuildWithDepsLogTest, TestInputMtimeRaceConditionWithDepFile) {
     state.Reset();
     EXPECT_TRUE(builder.AddTarget("out", &err));
     ASSERT_EQ("", err);
-    EXPECT_TRUE(builder.Build(&err));
+    EXPECT_EQ(builder.Build(&err), ExitSuccess);
     ASSERT_EQ(1u, command_runner_.commands_ran_.size());
 
     builder.command_runner_.release();
@@ -2930,7 +2931,7 @@ TEST_F(BuildTest, RestatDepfileDependency) {
   string err;
   EXPECT_TRUE(builder_.AddTarget("out", &err));
   ASSERT_EQ("", err);
-  EXPECT_TRUE(builder_.Build(&err));
+  EXPECT_EQ(builder_.Build(&err), ExitSuccess);
   EXPECT_EQ("", err);
 }
 
@@ -2962,7 +2963,7 @@ TEST_F(BuildWithDepsLogTest, RestatDepfileDependencyDepsLog) {
     EXPECT_TRUE(builder.AddTarget("out", &err));
     ASSERT_EQ("", err);
     fs_.Create("in1.d", "out: header.h");
-    EXPECT_TRUE(builder.Build(&err));
+    EXPECT_EQ(builder.Build(&err), ExitSuccess);
     EXPECT_EQ("", err);
 
     deps_log.Close();
@@ -2988,7 +2989,7 @@ TEST_F(BuildWithDepsLogTest, RestatDepfileDependencyDepsLog) {
     command_runner_.commands_ran_.clear();
     EXPECT_TRUE(builder.AddTarget("out", &err));
     ASSERT_EQ("", err);
-    EXPECT_TRUE(builder.Build(&err));
+    EXPECT_EQ(builder.Build(&err), ExitSuccess);
     EXPECT_EQ("", err);
 
     // Rule "true" should have run again, but the build of "out" should have
@@ -3021,7 +3022,7 @@ TEST_F(BuildWithDepsLogTest, DepFileOKDepsLog) {
     EXPECT_TRUE(builder.AddTarget("fo o.o", &err));
     ASSERT_EQ("", err);
     fs_.Create("fo o.o.d", "fo\\ o.o: blah.h bar.h\n");
-    EXPECT_TRUE(builder.Build(&err));
+    EXPECT_EQ(builder.Build(&err), ExitSuccess);
     EXPECT_EQ("", err);
 
     deps_log.Close();
@@ -3092,7 +3093,7 @@ TEST_F(BuildWithDepsLogTest, DiscoveredDepDuringBuildChanged) {
     EXPECT_TRUE(builder.AddTarget("out2", &err));
     EXPECT_FALSE(builder.AlreadyUpToDate());
 
-    EXPECT_TRUE(builder.Build(&err));
+    EXPECT_EQ(builder.Build(&err), ExitSuccess);
     EXPECT_TRUE(builder.AlreadyUpToDate());
 
     deps_log.Close();
@@ -3116,7 +3117,7 @@ TEST_F(BuildWithDepsLogTest, DiscoveredDepDuringBuildChanged) {
     EXPECT_TRUE(builder.AddTarget("out2", &err));
     EXPECT_FALSE(builder.AlreadyUpToDate());
 
-    EXPECT_TRUE(builder.Build(&err));
+    EXPECT_EQ(builder.Build(&err), ExitSuccess);
     EXPECT_TRUE(builder.AlreadyUpToDate());
 
     deps_log.Close();
@@ -3169,7 +3170,7 @@ TEST_F(BuildWithDepsLogTest, DepFileDepsLogCanonicalize) {
     // Note, different slashes from manifest.
     fs_.Create("a/b\\c\\d/e/fo o.o.d",
                "a\\b\\c\\d\\e\\fo\\ o.o: blah.h bar.h\n");
-    EXPECT_TRUE(builder.Build(&err));
+    EXPECT_EQ(builder.Build(&err), ExitSuccess);
     EXPECT_EQ("", err);
 
     deps_log.Close();
@@ -3320,7 +3321,7 @@ TEST_F(BuildTest, Console) {
   string err;
   EXPECT_TRUE(builder_.AddTarget("cons", &err));
   ASSERT_EQ("", err);
-  EXPECT_TRUE(builder_.Build(&err));
+  EXPECT_EQ(builder_.Build(&err), ExitSuccess);
   EXPECT_EQ("", err);
   ASSERT_EQ(1u, command_runner_.commands_ran_.size());
 }
@@ -3361,7 +3362,7 @@ TEST_F(BuildTest, DyndepReadyImplicitConnection) {
   string err;
   EXPECT_TRUE(builder_.AddTarget("out", &err));
   ASSERT_EQ("", err);
-  EXPECT_TRUE(builder_.Build(&err));
+  EXPECT_EQ(builder_.Build(&err), ExitSuccess);
   EXPECT_EQ("", err);
   ASSERT_EQ(2u, command_runner_.commands_ran_.size());
   EXPECT_EQ("touch tmp tmp.imp", command_runner_.commands_ran_[0]);
@@ -3428,7 +3429,7 @@ TEST_F(BuildTest, DyndepBuild) {
   EXPECT_EQ("", err);
 
   size_t files_created = fs_.files_created_.size();
-  EXPECT_TRUE(builder_.Build(&err));
+  EXPECT_EQ(builder_.Build(&err), ExitSuccess);
   EXPECT_EQ("", err);
 
   ASSERT_EQ(2u, command_runner_.commands_ran_.size());
@@ -3491,7 +3492,7 @@ TEST_F(BuildTest, DyndepBuildUnrelatedOutput) {
   EXPECT_TRUE(builder_.AddTarget("out", &err));
   EXPECT_EQ("", err);
 
-  EXPECT_TRUE(builder_.Build(&err));
+  EXPECT_EQ(builder_.Build(&err), ExitSuccess);
   EXPECT_EQ("", err);
   ASSERT_EQ(3u, command_runner_.commands_ran_.size());
   EXPECT_EQ("cp dd-in dd", command_runner_.commands_ran_[0]);
@@ -3523,7 +3524,7 @@ TEST_F(BuildTest, DyndepBuildDiscoverNewOutput) {
   EXPECT_TRUE(builder_.AddTarget("out", &err));
   EXPECT_EQ("", err);
 
-  EXPECT_TRUE(builder_.Build(&err));
+  EXPECT_EQ(builder_.Build(&err), ExitSuccess);
   EXPECT_EQ("", err);
   ASSERT_EQ(2u, command_runner_.commands_ran_.size());
   EXPECT_EQ("cp dd-in dd", command_runner_.commands_ran_[0]);
@@ -3625,7 +3626,7 @@ TEST_F(BuildTest, DyndepBuildDiscoverNewInput) {
   EXPECT_TRUE(builder_.AddTarget("out", &err));
   EXPECT_EQ("", err);
 
-  EXPECT_TRUE(builder_.Build(&err));
+  EXPECT_EQ(builder_.Build(&err), ExitSuccess);
   EXPECT_EQ("", err);
   ASSERT_EQ(3u, command_runner_.commands_ran_.size());
   EXPECT_EQ("cp dd-in dd", command_runner_.commands_ran_[0]);
@@ -3685,7 +3686,7 @@ TEST_F(BuildTest, DyndepBuildDiscoverNewInputWithTransitiveValidation) {
   EXPECT_TRUE(builder_.AddTarget("out", &err));
   EXPECT_EQ("", err);
 
-  EXPECT_TRUE(builder_.Build(&err));
+  EXPECT_EQ(builder_.Build(&err), ExitSuccess);
   EXPECT_EQ("", err);
   ASSERT_EQ(4u, command_runner_.commands_ran_.size());
   EXPECT_EQ("cp dd-in dd", command_runner_.commands_ran_[0]);
@@ -3718,7 +3719,7 @@ TEST_F(BuildTest, DyndepBuildDiscoverImplicitConnection) {
   string err;
   EXPECT_TRUE(builder_.AddTarget("out", &err));
   ASSERT_EQ("", err);
-  EXPECT_TRUE(builder_.Build(&err));
+  EXPECT_EQ(builder_.Build(&err), ExitSuccess);
   EXPECT_EQ("", err);
   ASSERT_EQ(3u, command_runner_.commands_ran_.size());
   EXPECT_EQ("cp dd-in dd", command_runner_.commands_ran_[0]);
@@ -3754,7 +3755,7 @@ TEST_F(BuildTest, DyndepBuildDiscoverOutputAndDepfileInput) {
   // Loading the depfile did not give tmp.imp a phony input edge.
   ASSERT_FALSE(GetNode("tmp.imp")->in_edge());
 
-  EXPECT_TRUE(builder_.Build(&err));
+  EXPECT_EQ(builder_.Build(&err), ExitSuccess);
   EXPECT_EQ("", err);
 
   // Loading the dyndep file gave tmp.imp a real input edge.
@@ -3793,7 +3794,7 @@ TEST_F(BuildTest, DyndepBuildDiscoverNowWantEdge) {
   string err;
   EXPECT_TRUE(builder_.AddTarget("out", &err));
   ASSERT_EQ("", err);
-  EXPECT_TRUE(builder_.Build(&err));
+  EXPECT_EQ(builder_.Build(&err), ExitSuccess);
   EXPECT_EQ("", err);
   ASSERT_EQ(3u, command_runner_.commands_ran_.size());
   EXPECT_EQ("cp dd-in dd", command_runner_.commands_ran_[0]);
@@ -3824,7 +3825,7 @@ TEST_F(BuildTest, DyndepBuildDiscoverNowWantEdgeAndDependent) {
   string err;
   EXPECT_TRUE(builder_.AddTarget("out", &err));
   ASSERT_EQ("", err);
-  EXPECT_TRUE(builder_.Build(&err));
+  EXPECT_EQ(builder_.Build(&err), ExitSuccess);
   EXPECT_EQ("", err);
   ASSERT_EQ(3u, command_runner_.commands_ran_.size());
   EXPECT_EQ("cp dd-in dd", command_runner_.commands_ran_[0]);
@@ -3895,7 +3896,7 @@ TEST_F(BuildWithLogTest, DyndepBuildDiscoverRestat) {
   string err;
   EXPECT_TRUE(builder_.AddTarget("out2", &err));
   ASSERT_EQ("", err);
-  EXPECT_TRUE(builder_.Build(&err));
+  EXPECT_EQ(builder_.Build(&err), ExitSuccess);
   ASSERT_EQ("", err);
   ASSERT_EQ(3u, command_runner_.commands_ran_.size());
   EXPECT_EQ("cp dd-in dd", command_runner_.commands_ran_[0]);
@@ -3911,7 +3912,7 @@ TEST_F(BuildWithLogTest, DyndepBuildDiscoverRestat) {
   // touch "out1", we should cancel the build of "out2".
   EXPECT_TRUE(builder_.AddTarget("out2", &err));
   ASSERT_EQ("", err);
-  EXPECT_TRUE(builder_.Build(&err));
+  EXPECT_EQ(builder_.Build(&err), ExitSuccess);
   ASSERT_EQ(1u, command_runner_.commands_ran_.size());
   EXPECT_EQ("true", command_runner_.commands_ran_[0]);
 }
@@ -3952,7 +3953,7 @@ TEST_F(BuildTest, DyndepBuildDiscoverScheduledEdge) {
   EXPECT_TRUE(builder_.AddTarget("out1", &err));
   EXPECT_TRUE(builder_.AddTarget("out2", &err));
   ASSERT_EQ("", err);
-  EXPECT_TRUE(builder_.Build(&err));
+  EXPECT_EQ(builder_.Build(&err), ExitSuccess);
   EXPECT_EQ("", err);
   ASSERT_EQ(3u, command_runner_.commands_ran_.size());
   // Depending on how the pointers in Plan::ready_ work out, the first
@@ -4003,7 +4004,7 @@ TEST_F(BuildTest, DyndepTwoLevelDirect) {
   string err;
   EXPECT_TRUE(builder_.AddTarget("out2", &err));
   ASSERT_EQ("", err);
-  EXPECT_TRUE(builder_.Build(&err));
+  EXPECT_EQ(builder_.Build(&err), ExitSuccess);
   EXPECT_EQ("", err);
   ASSERT_EQ(3u, command_runner_.commands_ran_.size());
   EXPECT_EQ("cp dd1-in dd1", command_runner_.commands_ran_[0]);
@@ -4048,7 +4049,7 @@ TEST_F(BuildTest, DyndepTwoLevelIndirect) {
   string err;
   EXPECT_TRUE(builder_.AddTarget("out2", &err));
   ASSERT_EQ("", err);
-  EXPECT_TRUE(builder_.Build(&err));
+  EXPECT_EQ(builder_.Build(&err), ExitSuccess);
   EXPECT_EQ("", err);
   ASSERT_EQ(3u, command_runner_.commands_ran_.size());
   EXPECT_EQ("cp dd1-in dd1", command_runner_.commands_ran_[0]);
@@ -4088,7 +4089,7 @@ TEST_F(BuildTest, DyndepTwoLevelDiscoveredReady) {
   EXPECT_TRUE(builder_.AddTarget("out", &err));
   EXPECT_EQ("", err);
 
-  EXPECT_TRUE(builder_.Build(&err));
+  EXPECT_EQ(builder_.Build(&err), ExitSuccess);
   EXPECT_EQ("", err);
   ASSERT_EQ(4u, command_runner_.commands_ran_.size());
   EXPECT_EQ("cp dd1-in dd1", command_runner_.commands_ran_[0]);
@@ -4128,7 +4129,7 @@ TEST_F(BuildTest, DyndepTwoLevelDiscoveredDirty) {
   EXPECT_TRUE(builder_.AddTarget("out", &err));
   EXPECT_EQ("", err);
 
-  EXPECT_TRUE(builder_.Build(&err));
+  EXPECT_EQ(builder_.Build(&err), ExitSuccess);
   EXPECT_EQ("", err);
   ASSERT_EQ(5u, command_runner_.commands_ran_.size());
   EXPECT_EQ("cp dd1-in dd1", command_runner_.commands_ran_[0]);
@@ -4150,7 +4151,7 @@ TEST_F(BuildTest, Validation) {
   EXPECT_TRUE(builder_.AddTarget("out", &err));
   EXPECT_EQ("", err);
 
-  EXPECT_TRUE(builder_.Build(&err));
+  EXPECT_EQ(builder_.Build(&err), ExitSuccess);
   EXPECT_EQ("", err);
 
   EXPECT_EQ(2u, command_runner_.commands_ran_.size());
@@ -4166,7 +4167,7 @@ TEST_F(BuildTest, Validation) {
   EXPECT_TRUE(builder_.AddTarget("out", &err));
   ASSERT_EQ("", err);
 
-  EXPECT_TRUE(builder_.Build(&err));
+  EXPECT_EQ(builder_.Build(&err), ExitSuccess);
   EXPECT_EQ("", err);
 
   ASSERT_EQ(1u, command_runner_.commands_ran_.size());
@@ -4183,7 +4184,7 @@ TEST_F(BuildTest, Validation) {
   EXPECT_TRUE(builder_.AddTarget("out", &err));
   ASSERT_EQ("", err);
 
-  EXPECT_TRUE(builder_.Build(&err));
+  EXPECT_EQ(builder_.Build(&err), ExitSuccess);
   EXPECT_EQ("", err);
 
   ASSERT_EQ(1u, command_runner_.commands_ran_.size());
@@ -4202,7 +4203,7 @@ TEST_F(BuildTest, ValidationDependsOnOutput) {
   EXPECT_TRUE(builder_.AddTarget("out", &err));
   EXPECT_EQ("", err);
 
-  EXPECT_TRUE(builder_.Build(&err));
+  EXPECT_EQ(builder_.Build(&err), ExitSuccess);
   EXPECT_EQ("", err);
 
   EXPECT_EQ(2u, command_runner_.commands_ran_.size());
@@ -4217,7 +4218,7 @@ TEST_F(BuildTest, ValidationDependsOnOutput) {
   EXPECT_TRUE(builder_.AddTarget("out", &err));
   ASSERT_EQ("", err);
 
-  EXPECT_TRUE(builder_.Build(&err));
+  EXPECT_EQ(builder_.Build(&err), ExitSuccess);
   EXPECT_EQ("", err);
 
   EXPECT_EQ(2u, command_runner_.commands_ran_.size());
@@ -4233,7 +4234,7 @@ TEST_F(BuildTest, ValidationDependsOnOutput) {
   EXPECT_TRUE(builder_.AddTarget("out", &err));
   ASSERT_EQ("", err);
 
-  EXPECT_TRUE(builder_.Build(&err));
+  EXPECT_EQ(builder_.Build(&err), ExitSuccess);
   EXPECT_EQ("", err);
 
   ASSERT_EQ(1u, command_runner_.commands_ran_.size());
@@ -4270,7 +4271,7 @@ TEST_F(BuildWithDepsLogTest, ValidationThroughDepfile) {
     EXPECT_TRUE(builder.AddTarget("out2", &err));
     ASSERT_EQ("", err);
 
-    EXPECT_TRUE(builder.Build(&err));
+    EXPECT_EQ(builder.Build(&err), ExitSuccess);
     EXPECT_EQ("", err);
 
     // On the first build, only the out2 command is run.
@@ -4306,7 +4307,7 @@ TEST_F(BuildWithDepsLogTest, ValidationThroughDepfile) {
     EXPECT_TRUE(builder.AddTarget("out2", &err));
     ASSERT_EQ("", err);
 
-    EXPECT_TRUE(builder.Build(&err));
+    EXPECT_EQ(builder.Build(&err), ExitSuccess);
     EXPECT_EQ("", err);
 
     // The out and validate actions should have been run as well as out2.
@@ -4331,7 +4332,7 @@ TEST_F(BuildTest, ValidationCircular) {
   EXPECT_TRUE(builder_.AddTarget("out", &err));
   EXPECT_EQ("", err);
 
-  EXPECT_TRUE(builder_.Build(&err));
+  EXPECT_EQ(builder_.Build(&err), ExitSuccess);
   EXPECT_EQ("", err);
 
   EXPECT_EQ(2u, command_runner_.commands_ran_.size());
@@ -4346,7 +4347,7 @@ TEST_F(BuildTest, ValidationCircular) {
   EXPECT_TRUE(builder_.AddTarget("out", &err));
   ASSERT_EQ("", err);
 
-  EXPECT_TRUE(builder_.Build(&err));
+  EXPECT_EQ(builder_.Build(&err), ExitSuccess);
   EXPECT_EQ("", err);
 
   ASSERT_EQ(1u, command_runner_.commands_ran_.size());
@@ -4362,7 +4363,7 @@ TEST_F(BuildTest, ValidationCircular) {
   EXPECT_TRUE(builder_.AddTarget("out", &err));
   ASSERT_EQ("", err);
 
-  EXPECT_TRUE(builder_.Build(&err));
+  EXPECT_EQ(builder_.Build(&err), ExitSuccess);
   EXPECT_EQ("", err);
 
   ASSERT_EQ(1u, command_runner_.commands_ran_.size());

--- a/src/exit_status.h
+++ b/src/exit_status.h
@@ -15,10 +15,19 @@
 #ifndef NINJA_EXIT_STATUS_H_
 #define NINJA_EXIT_STATUS_H_
 
-enum ExitStatus {
-  ExitSuccess,
+// The underlying type of the ExitStatus enum, used to represent a platform-specific
+// process exit code.
+#ifdef _WIN32
+#define EXIT_STATUS_TYPE unsigned long
+#else  // !_WIN32
+#define EXIT_STATUS_TYPE int
+#endif  // !_WIN32
+
+
+enum ExitStatus : EXIT_STATUS_TYPE {
+  ExitSuccess=0,
   ExitFailure,
-  ExitInterrupted
+  ExitInterrupted=130,
 };
 
 #endif  // NINJA_EXIT_STATUS_H_

--- a/src/ninja.cc
+++ b/src/ninja.cc
@@ -42,8 +42,8 @@
 #include "clean.h"
 #include "command_collector.h"
 #include "debug_flags.h"
-#include "depfile_parser.h"
 #include "disk_interface.h"
+#include "exit_status.h"
 #include "graph.h"
 #include "graphviz.h"
 #include "json.h"
@@ -165,7 +165,7 @@ struct NinjaMain : public BuildLogUser {
 
   /// Build the targets listed on the command line.
   /// @return an exit code.
-  int RunBuild(int argc, char** argv, Status* status);
+  ExitStatus RunBuild(int argc, char** argv, Status* status);
 
   /// Dump the output requested by '-d stats'.
   void DumpMetrics();
@@ -281,7 +281,7 @@ bool NinjaMain::RebuildManifest(const char* input_file, string* err,
   if (builder.AlreadyUpToDate())
     return false;  // Not an error, but we didn't rebuild.
 
-  if (!builder.Build(err))
+  if (builder.Build(err) != ExitSuccess)
     return false;
 
   // The manifest was only rebuilt if it is now dirty (it may have been cleaned
@@ -1541,12 +1541,12 @@ bool NinjaMain::EnsureBuildDirExists() {
   return true;
 }
 
-int NinjaMain::RunBuild(int argc, char** argv, Status* status) {
+ExitStatus NinjaMain::RunBuild(int argc, char** argv, Status* status) {
   string err;
   vector<Node*> targets;
   if (!CollectTargetsFromArgs(argc, argv, &targets, &err)) {
     status->Error("%s", err.c_str());
-    return 1;
+    return ExitFailure;
   }
 
   disk_interface_.AllowStatCache(g_experimental_statcache);
@@ -1557,7 +1557,7 @@ int NinjaMain::RunBuild(int argc, char** argv, Status* status) {
     if (!builder.AddTarget(targets[i], &err)) {
       if (!err.empty()) {
         status->Error("%s", err.c_str());
-        return 1;
+        return ExitFailure;
       } else {
         // Added a target that is already up-to-date; not really
         // an error.
@@ -1572,18 +1572,18 @@ int NinjaMain::RunBuild(int argc, char** argv, Status* status) {
     if (config_.verbosity != BuildConfig::NO_STATUS_UPDATE) {
       status->Info("no work to do.");
     }
-    return 0;
+    return ExitSuccess;
   }
 
-  if (!builder.Build(&err)) {
+  ExitStatus exit_status = builder.Build(&err);
+  if (exit_status != ExitSuccess) {
     status->Info("build stopped: %s.", err.c_str());
     if (err.find("interrupted by user") != string::npos) {
-      return 130;
+      return ExitInterrupted;
     }
-    return 1;
   }
 
-  return 0;
+  return exit_status;
 }
 
 #ifdef _MSC_VER
@@ -1801,7 +1801,7 @@ NORETURN void real_main(int argc, char** argv) {
 
     ninja.ParsePreviousElapsedTimes();
 
-    int result = ninja.RunBuild(argc, argv, status);
+    ExitStatus result = ninja.RunBuild(argc, argv, status);
     if (g_metrics)
       ninja.DumpMetrics();
     exit(result);

--- a/src/status.h
+++ b/src/status.h
@@ -16,6 +16,7 @@
 #define NINJA_STATUS_H_
 
 #include <string>
+#include "exit_status.h"
 
 struct BuildConfig;
 struct Edge;
@@ -29,7 +30,7 @@ struct Status {
   virtual void BuildEdgeStarted(const Edge* edge,
                                 int64_t start_time_millis) = 0;
   virtual void BuildEdgeFinished(Edge* edge, int64_t start_time_millis,
-                                 int64_t end_time_millis, bool success,
+                                 int64_t end_time_millis, ExitStatus exit_code,
                                  const std::string& output) = 0;
   virtual void BuildStarted() = 0;
   virtual void BuildFinished() = 0;

--- a/src/status_printer.cc
+++ b/src/status_printer.cc
@@ -33,6 +33,7 @@
 
 #include "build.h"
 #include "debug_flags.h"
+#include "exit_status.h"
 
 using namespace std;
 
@@ -174,7 +175,7 @@ void StatusPrinter::RecalculateProgressPrediction() {
 }
 
 void StatusPrinter::BuildEdgeFinished(Edge* edge, int64_t start_time_millis,
-                                      int64_t end_time_millis, bool success,
+                                      int64_t end_time_millis, ExitStatus exit_code,
                                       const string& output) {
   time_millis_ = end_time_millis;
   ++finished_edges_;
@@ -202,16 +203,17 @@ void StatusPrinter::BuildEdgeFinished(Edge* edge, int64_t start_time_millis,
   --running_edges_;
 
   // Print the command that is spewing before printing its output.
-  if (!success) {
+  if (exit_code != ExitSuccess) {
     string outputs;
     for (vector<Node*>::const_iterator o = edge->outputs_.begin();
          o != edge->outputs_.end(); ++o)
       outputs += (*o)->path() + " ";
 
+    string failed = "FAILED: [code=" + std::to_string(exit_code) + "] ";
     if (printer_.supports_color()) {
-        printer_.PrintOnNewLine("\x1B[31m" "FAILED: " "\x1B[0m" + outputs + "\n");
+        printer_.PrintOnNewLine("\x1B[31m" + failed + "\x1B[0m" + outputs + "\n");
     } else {
-        printer_.PrintOnNewLine("FAILED: " + outputs + "\n");
+        printer_.PrintOnNewLine(failed + outputs + "\n");
     }
     printer_.PrintOnNewLine(edge->EvaluateCommand() + "\n");
   }

--- a/src/status_printer.h
+++ b/src/status_printer.h
@@ -16,6 +16,7 @@
 #include <cstdint>
 #include <queue>
 
+#include "exit_status.h"
 #include "explanations.h"
 #include "line_printer.h"
 #include "status.h"
@@ -31,7 +32,7 @@ struct StatusPrinter : Status {
 
   void BuildEdgeStarted(const Edge* edge, int64_t start_time_millis) override;
   void BuildEdgeFinished(Edge* edge, int64_t start_time_millis,
-                                 int64_t end_time_millis, bool success,
+                                 int64_t end_time_millis, ExitStatus exit_code,
                                  const std::string& output) override;
   void BuildStarted() override;
   void BuildFinished() override;

--- a/src/subprocess-posix.cc
+++ b/src/subprocess-posix.cc
@@ -12,6 +12,7 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+#include "exit_status.h"
 #include "subprocess.h"
 
 #include <sys/select.h>
@@ -165,15 +166,16 @@ ExitStatus Subprocess::Finish() {
 #endif
 
   if (WIFEXITED(status)) {
-    int exit = WEXITSTATUS(status);
-    if (exit == 0)
-      return ExitSuccess;
-  } else if (WIFSIGNALED(status)) {
+    // propagate the status transparently
+    return static_cast<ExitStatus>(WEXITSTATUS(status));
+  }
+  if (WIFSIGNALED(status)) {
     if (WTERMSIG(status) == SIGINT || WTERMSIG(status) == SIGTERM
         || WTERMSIG(status) == SIGHUP)
       return ExitInterrupted;
   }
-  return ExitFailure;
+  // At this point, we exit with any other signal+128
+  return static_cast<ExitStatus>(status + 128);
 }
 
 bool Subprocess::Done() const {

--- a/src/subprocess-win32.cc
+++ b/src/subprocess-win32.cc
@@ -12,6 +12,7 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+#include "exit_status.h"
 #include "subprocess.h"
 
 #include <assert.h>
@@ -198,9 +199,8 @@ ExitStatus Subprocess::Finish() {
   CloseHandle(child_);
   child_ = NULL;
 
-  return exit_code == 0              ? ExitSuccess :
-         exit_code == CONTROL_C_EXIT ? ExitInterrupted :
-                                       ExitFailure;
+  return exit_code == CONTROL_C_EXIT ? ExitInterrupted :
+                                       static_cast<ExitStatus>(exit_code);
 }
 
 bool Subprocess::Done() const {

--- a/src/subprocess_test.cc
+++ b/src/subprocess_test.cc
@@ -14,6 +14,7 @@
 
 #include "subprocess.h"
 
+#include "exit_status.h"
 #include "test.h"
 
 #ifndef _WIN32
@@ -50,7 +51,8 @@ TEST_F(SubprocessTest, BadCommandStderr) {
     subprocs_.DoWork();
   }
 
-  EXPECT_EQ(ExitFailure, subproc->Finish());
+  ExitStatus exit = subproc->Finish();
+  EXPECT_NE(ExitSuccess, exit);
   EXPECT_NE("", subproc->GetOutput());
 }
 
@@ -64,7 +66,8 @@ TEST_F(SubprocessTest, NoSuchCommand) {
     subprocs_.DoWork();
   }
 
-  EXPECT_EQ(ExitFailure, subproc->Finish());
+  ExitStatus exit = subproc->Finish();
+  EXPECT_NE(ExitSuccess, exit);
   EXPECT_NE("", subproc->GetOutput());
 #ifdef _WIN32
   ASSERT_EQ("CreateProcess failed: The system cannot find the file "

